aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -103,20 +109,19 @@ Tensor sparse_csr_tensor(
       options.layout());
   TORCH_CHECK(crow_indices.numel() >= 1, "expected crow_indices.numel() >= 1, but got ",
               crow_indices.numel());
-  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
   std::array<int64_t, 2> size;
-
   if (col_indices.numel() > 0) {
     size[0] = crow_indices.numel() - 1;
-    Tensor max_col_indices = std::get<0>(col_indices.max(0, false));
+    Tensor max_col_indices = std::get<0>(col_indices.max(0, false)).to(kCPU);
```

Comment:
SYNC!!! You sure about this?

Reply:
Umhhh let me review more about CUDA SYNC. I know that I will learn something interesting in the process.

Reply:
This is what COO tensor constructor does as well.

When you ask this, @ezyang , the question arises: should we trust the value-content of `crow_indices` and `col_indices` (or `indices` in the case of COO) provided by users/callers, and not perform these value-based integrity checks?

If yes, then it will speed up the constructions of sparse tensors (especially those using cuda storage) but there is a risk that invalid data will crash python at some point. For example, crow_indices must be an increasing sequence and currently, we don't have a test for it. Here's an example that will crash python because of this:
```
torch.sparse_csr_tensor(torch.tensor([0, 40, 3]), torch.tensor([2, 0, 1]), [1, 2, 3]).to_dense()
```
(the crash happens in `to_dense` call and could be avoided there but it is not really `to_dense` fault that CSR tensor contains invalid indices data, IMHO).

If no, what would be the recommended way to perform value-based integrity checks when using CUDA storage?

Possible options:
1. have unsafe tensor constructors that skip the value-based integrity checks; to be used only when the caller ensures that integrity conditions are met
2. perform value-based integrity checks only when using CPU storage

Reply:
(1) is typically a good idea, because when you are writing kernels you typically want to assume the integrity checks are upheld (otherwise the algorithm is just wrong) and if you force the integrity checks always you will effectively sync on every kernel call.

Another general principle which is useful to think about is that we are willing to pay some performance cost to do error checking in kernels, but usually this manifests as willingness to add CUDA asserts (e.g., when we do embedding lookups, we will CUDA assert that they are in bounds). CUDA asserts are less expensive than syncing and might be a good way of doing integrity checks in other situations.

cc @ngimel and @fmassa for other opinions

Reply:
I added  an `unsafe csr tensor constructors` that skip the value-based integrity checks.
Still need to double-check indices invariants implementation.   See [CSR tensor invariants](https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md)
 
cc @pearu